design/velero-client-download-apiserver_design.md

```diff
@@ -0,0 +1,73 @@
+# Velero Endpoint Download Client
+
+## Abstract
+
+<!-- One to two sentences that describes the goal of this proposal and the problem being solved by the proposed change.
+The reader should be able to tell by the title, and the opening paragraph, if this document is relevant to them. -->
+Velero data is stored in an object store.
+When CLI users want to download data from the object store, it connects to the specific object store directly to download data.
+This assumes the target storing the data is always accessible from the client side. This is not always true, especially in the on-premise environments.
+Velero may also save the data in non object stores in the future so an object store may not be available for the CLI to download from.
+
+This design proposes to add a new endpoint on the Velero server that CLI users can connect to and download data from without having to connect to an object store directly.
+
+## Background
+Velero CLI commands such as following download data from the object store directly today.
+```
+velero backup download
+velero describe
+velero backup logs
+...
+```
+We want them to work even when the object store is not accessible from the client side.
+Even if the object store is accessible from the client side, today things like cacert and insecure-skip-tls-verify flags are specified manually from the client even though the Velero server already has the information. This work eliminates the client from having to know about the object store configuration and certificates.
+## Goals
+- A short list of things which will be accomplished by implementing this proposal.
+- Two things is ok.
+- Three is pushing it.
+- More than three goals suggests that the proposal's scope is too large.
+
+## Non Goals
+- A short list of items which are:
+- a. out of scope
+- b. follow on items which are deliberately excluded from this proposal.
+
+
+## High-Level Design
+One to two paragraphs that describe the high level changes that will be made to implement this proposal.
+
+We will [configure a Kubernetes API Aggregation Layer](https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer/) to communicate with a new Velero extension apiserver in order to authenticate and authorize requests to download data from the Velero server.
```

Comment:
I have been cautioned around using this.

An alternative is to have an Ingress reachable by velero CLI to query.

Reply:
We can utilize https://dev-k8sref-io.web.app/docs/authentication/tokenreview-v1/
https://dev-k8sref-io.web.app/docs/authorization/subjectaccessreview-v1/ for RBAC of the authorized velero user.

Moving to using Ingress to a velero pod that can redirect requests.

Reply:
We can reuse DownloadRequest CR and instead of providing object storage url to an object, this API can retrieve the file from storage location then serve the file to the user.

Reply:
So instead of downloadrequest CR status having URL to aws/gcp/azure, it will be url to velero ingress reachable by cli.

Reply:
How about listing the pros and cons for both the aggregation and ingress approaches and making comparisons?

I think of some cons of the ingress approach:
- Need to handle auth by ourselves
- Introduce different styles of API: currently, all Velero APIs are k8s style; the ingress approach introduces different style of APIs
- Besides the ingress, we also need to support some other kinds of expose methods depend on the capabilities of the k8s: NodePort, ClusterIP, LoadBalancer. This introduces more configuration items when installing Velero. e.g. the ingress is different for different ingress controllers, so users need to set the ingress controller type
- Introduce some external dependencies: ingress needs the ingress controller ready and the LoadBalancer needs the load balancer provider ready

Reply:
One big issue for the aggregation approach I can think of is how to support installing multiple Velero in one cluster.

Reply:
The aggregation approach also requires [flags](https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer/#enable-kubernetes-apiserver-flags) to be enabled in the apiserver.

Reply:
Just to be clear there is no reason to not follow normal k8s api structure for this, unless you want it to behave like a sub-resource like /logs or /exec even those follow k8s apis.

Doing auth yourself, but you would have had to add similar calls in the middleware of the aggregation api server because you want to check more SAR checks then just the basic ones for the resource being requested

You can still use RBAC for the resource like this.

With that said it does very much make it easier to run multiple veleros because you will have these servers tied to a particular install like you mentioned.

You should also note that many of the Aggregated API libraries and tools are alpha and have not had updates since 2022 AFAICT.

Reply:
We could add the `namespace` to the `group` of `API Service` to support multiple Velero installations in one cluster. The API looks like `/apis/namespace01.velero.io/v1alpha1/...` for the Velero installed in the namespace `namespace01`.

We could choose the ingress approach, it can work absolutely. But my big concern is that current Velero exposes services based on the Kubernetes mechanism(CRD), Kubernetes handles the auth/TLS cert/etc. stuff. If we expose services via ingress/LB, we need to handle all of these by ourselves and two kinds of mechanisms exist for Velero.

Reply:
Another idea came up which is to use kubectl cp to copy file off of a velero pod. This way, no ingress will be needed. Velero pod downloads the file to a path, and kubectl cp then velero pod deletes file.

Reply:
I am little bit confused by "auth" requirement for the download APIs. No other Velero APIs require any authentication/RBAC other than Kubernetes auth/RBAC. Why  would download APIs be any different? Just making sure I am not missing something.

Reply:
> No other Velero APIs require any authentication/RBAC other than Kubernetes auth/RBAC.

All prior Velero APIs are implemeted as CustomResourceDefinitions. RBAC for these can be defined in [cluster]roles which define who can create what CRs where.

With API Aggregation layer, or ingress, I believe this falls outside of the general RBAC controls available to core k8s APIs and extensions via CustomResourceDefinitions.

See a section about [Extension Apiserver Authorizes the Request](https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer/#extension-apiserver-authorizes-the-request) which mentions the need for the apiserver to perform `SubjectAccessReview`. While an API Aggregation Layer might already require requests to contain API Group or GVK of sorts (I need to do more reading) which could be controlled by Role/ClusterRole, the Ingress method is very much open ended.

The Ingress method would require the client request to specifically include user/token information to be able to validate who they are and if they are allowed to do a certain action, and is not going to have GroupVersionKind that is definable in Role/ClusterRole. So velero ingress pod would have to explicitly does SubjectAccessReview and [TokenReview](https://dev-k8sref-io.web.app/docs/authentication/tokenreview-v1/) by ourselves.

Reply:
Ingress solution without auth consideration is akin to having an ingress to nginx in the cluster serving a website without auth. Kube by default doesn't care/know who or enforce what's being requested from the website server exposed via Ingress.

Reply:
Can we add this discussion to a community meeting, so that we can discuss this more syncrounously. 

@ywk253100 your concern regarding the need for auth/z is not an un-founded one, but one that I don't think Aggregated API's helps you with as much as you expect.

In both cases you will need to:
1. Set up an ingress/service
2. Create some certificate exchange. (for aggregated API, it is a custom CA, for your own service, it is a CA that the clients can trust, either using ingress re-encrypt from the cluster, or pass-through etc.)
3. Aggregated API does contain the middleware that does the TokenReview; you can still use this, though, by setting up this: https://github.com/kubernetes/apiserver/blob/master/pkg/authentication/authenticatorfactory/delegating.go#L68 in a middleware. (This is all that apiserver-runtime does)
4. Same basic process for setting up authorization, but that only focuses on the API that is being called. IIUC we need to make sure that the user has access to a different resource, therefore we would have to add a SAR check anyway even if we used apiserver-runtime. 

With the above said, creating the server in whatever way is orthogonal from making the Kubernetes API Server aware and proxy vs hitting the server directly. I hope this proves that regardless of mechanism you have to basically solve both cases.

Reply:
> Can we add this discussion to a community meeting, so that we can discuss this more syncrounously.
Sure. Let's discuss this after the Chinese New Year holiday as some of us are going to take vacations from this week.

Firstly, let's clarify whether the service provided by the download server is API. If it is, I think the aggregation is better. 
Let me explain my point about this more clearly: I am not very concerned about the implementation details of the auth/cert etc. stuff, both aggregation and ingress approaches should work. What I'm concerned about is that if we choose the ingress approach, there will be two mechanisms for Velero APIs: one is via the kube API server with CRD, and the other one is via ingress. All these APIs are provided by Velero, but they are not unified, they are even exposed by two different servers. If the third parties want to integrate Velero, they need to connect to two servers to all the APIs.

If we don't think it is an API, I'm fine for both approaches.

Reply:
Just so that I am understanding this clearly, we are considering ingress/aggregate API approach (instead of CRDs) because we are talking about streaming data (similar to kubectl logs)?

Another approach I can think of is to create a new CRD based API to initiate the download. Velero can then create the necessary files inside the pod at some location and the user can copy them using "kubectl cp". This sounds a bit hackish but will completely bypass any new API mechanism.

Reply:
> Just so that I am understanding this clearly, we are considering ingress/aggregate API approach (instead of CRDs) because we are talking about streaming data (similar to kubectl logs)?

Yes. Currently, Velero CLI gets a direct access URL of the object store and downloads the logs/backup tar/etc. from the object store directly, but the object store isn't always accessible by the CLI, so we want to change the way to do the download work

> Another approach I can think of is to create a new CRD based API to initiate the download. Velero can then create the necessary files inside the pod at some location and the user can copy them using "kubectl cp". This sounds a bit hackish but will completely bypass any new API mechanism.

This doesn't work because `kubectl cp` requires the `tar` binary to be present in the container while the Velero container doesn't contain it.

Reply:
@mpryc also suggested `kubectl cp` to me awhile back.

> requires the tar binary to be present in the container

that sounds easy to fix.

Also if we consider kubectl cp, we get rid of the rbac implementation, as you probably can only run cp on containers you have access to. If you have access to velero container, you can already check for secrets written to credentialsFile inside velero pod.

Reply:
It would only break apart if there's not enough disk space on the node to download volume backups.

Reply:
That is true though Velero can always accept a PVC for downloading logs, still keeping default as using node storage ("emptyDir").

Reply:
For `kubectl cp` We can use [range in HTTP header in s3 request](https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html#AmazonS3-GetObject-request-header-Range:~:text=of%20an%20object.-,Range,-Downloads%20the%20specified) to break big files into chunks then reassemble client side.
> Downloads the specified byte range of an object.

Reply:
Our product at Replicated uses the `github.com/vmware-tanzu/velero/pkg/cmd/util/downloadrequest` package to [download/stream backup logs](https://github.com/replicatedhq/kots/blob/30eeb4b70f6fa7c56aeb85bcb827e642facb5c9b/pkg/kotsadmsnapshot/download.go#L85-L106). It would be great to ensure backwards compatibility here in order to avoid breaking our customers' installations when they upgrade Velero.

Reply:
I honestly don't see a difference between, exposing a pod to CP (user will need exec on the pod then right)? And exposing a service with/ actual RBAC checks for the backup, etc. I also think that keeping the old behavior, such that the provider has a way for the user to pull the data directly, would be a good thing.

Reply:
After discussing with @weshayutin I think we can keep backwards compatibility without introducing auth headers requirement by essentially recreating the presigned URL behavior
> [You can use presigned URLs to grant time-limited access to objects](https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-presigned-url.html#:~:text=You%20can%20use%20presigned%20URLs%20to%20grant%20time%2Dlimited%20access%20to%20objects)

Reply:
downloadUrl for each download request will contain uuid and expire shortly.

Reply:
presigned would be optional, we should auth by default using subjectaccessreviews and tokenreview